### DISCUSSION TITLE
How to override C# class in JS and return instance back to C#

### DISCUSSION THREAD

**Discussion by achimmihca** [sebastienros/jint#1429] [Q&A]
Is it possible to
- pass a C# class to JS (yes)
- extend it in JS (yes)
    - and override methods (not sure)
- and pass a new instance of the subclass back to C#? (not sure)

My goal is to integrate Jint in a Unity game to let users register custom code, e.g. a `HighscoreProvider`.

```
// C# class
public class HighscoreProvider
{
    public virtual int GetScore() { return 10; }
}
```

```
// JS subclass
class MyHighscoreProvider extends HighscoreProvider
{
    constructor() { super(); this.score = 2000; }

    // Override C# method
    getScore() { return this.score; }
}

log("this.score: " + new MyHighscoreProvider().score); output: "this.score: 2000"
log("getScore: " + new MyHighscoreProvider().getScore()); // output: "getScore: 10"
log("GetScore: " + new MyHighscoreProvider().GetScore()); // output: "GetScore: 10"
```

Is this possible?
In my test, it always returns the value of the C# method. Thus, overriding the method does not seem to work.

BTW: how is upper/lower case handled? I notice that PascalCase in C# seems to be available as camelCase in JS.

> pass a new instance of the subclass back to C#

I am thinking about a ScriptRegistry that is accessible in JS so users can add stuff to it.
```
// C#
public class ScriptRegistry
{
    public List<HighscoreProvider> HighscoreProviders { get; private set; } = new();
    
    public void AddHighscoreProvider(HighscoreProvider highscoreProvider)
    {
        HighscoreProviders.Add(highscoreProvider);
    }
}
```

```
// JS
// add JS instance to C# ScriptRegistry
scriptRegistry.addHighscoreProvider(new MyHighscoreProvider())
```

The HighscoreProvider and Registry are made available to the Engine:
```
engine.SetValue(nameof(scriptRegistry), scriptRegistry);
engine.SetValue(nameof(HighscoreProvider), TypeReference.CreateTypeReference(engine, typeof(HighscoreProvider)));
```

Above nearly works. `AddHighscoreProvider` is called with some object, but it does not seem to include the overridden `getScore` method. `GetScore` still returns 10.

What would be even better than a C# class would be a C# interface that can be implemented in JS.
Is it possible to implement a C# interface in JS and return an instance to C#?

**Comment by Genteure** [comment]
- pass a C# class to JS (yes)
- extend it in JS (yes, but only available in JS)
  - and override methods (no, it's only available in JS)
- and pass a new instance of the subclass back to C#? (yes, but might not work as you imagined)

Instead of trying to implement C# interface directly in js (which doesn't work), you can make a C# class that implement your C# interface that calls the corresponding js function.

<details>
<summary>untested pseudocode</summary>

```csharp
public interface IHighscoreProvider
{
    void SetScore(int score);
    int GetScore();
}

public class ScriptingHighscoreProvider : IHighscoreProvider
{
    private Engine engine;

    public ScriptingHighscoreProvider()
    {
        this.engine = new Engine();
        var jscode = File.ReadAllText("...");
        this.engine.Execute(jscode);
    }

    public void SetScore(int score)
    {
        this.engine.Invoke("setScore", score);
    }

    public int GetScore()
    {
        return (int)this.engine.Invoke("getScore").AsNumber();
    }
}
```

</details>

Tested example:

```csharp
public interface IHighscoreProvider
{
    void SetScore(int score);
    int GetScore();
}

public class ScriptingHighscoreProvider : IHighscoreProvider
{
    private readonly Engine engine;
    private ObjectInstance provider;

    public ScriptingHighscoreProvider()
    {
        engine = new Engine();
        engine.SetValue("setHighscoreProvider", (JsValue value) =>
        {
            if (!value.IsObject())
            {
                throw new JavaScriptException(engine.Realm.Intrinsics.Error, "only object can be passed");
            }
            provider = value.AsObject();
        });

        engine.Execute("""
class MyHighscoreProvider {
    constructor() { this.score = 2000; }
    setScore(score) { this.score = score; }
    getScore() { return this.score; }
}

setHighscoreProvider(new MyHighscoreProvider())
""");

        if (provider is null)
        {
            throw new Exception("script did not call setHighscoreProvider");
        }
    }

    public int GetScore()
    {
        return (int) provider.Get("getScore").Call(thisObj: provider).AsNumber();
    }

    public void SetScore(int score)
    {
        provider.Get("setScore").Call(thisObj: provider, score);
    }
}

public class Test
{
    [Fact]
    public void TestRun()
    {
        IHighscoreProvider provider = new ScriptingHighscoreProvider();

        Assert.Equal(2000, provider.GetScore());
        provider.SetScore(123);
        Assert.Equal(123, provider.GetScore());
    }
}
```

Keep in mind `Engine` is not thread safe and should only be called in one thread at a time.

Depending on how you want to implement scripting in your project, you could also create a new `Engine` for each invocation and keep the states out of Jint.

**Reply by Genteure** [reply]
I have edited the answer with a tested runnable example.

**Reply by achimmihca** [reply]
Thanks!
